packages/headless/src/api/knowledge/knowledge-api-answer.ts

```diff
@@ -0,0 +1,177 @@
+import {
+  EventSourceMessage,
+  fetchEventSource,
+} from '@microsoft/fetch-event-source';
+import {GeneratedContentFormat} from '../../features/generated-answer/generated-response-format';
+import {selectQuery} from '../../features/query/query-selectors';
+import {QueryState} from '../../features/query/query-state';
+import {selectSearchHub} from '../../features/search-hub/search-hub-selectors';
+import {ConfigurationSection} from '../../state/state-sections';
+import {GeneratedAnswerCitation} from '../generated-answer/generated-answer-event-payload';
+import {SearchRequest} from '../search/search/search-request';
+import {knowledgeApi} from './knowledge-api';
+
+type StateNeededByKnowledgeAPI = ConfigurationSection;
+
+interface ErrorPayload {
+  message?: string;
+  code?: number;
+}
+
+class FatalError extends Error {
+  constructor(public payload: ErrorPayload) {
+    super(payload.message);
+  }
+}
+
+interface GeneratedAnswerStream {
+  answerStyle: string | undefined;
+  contentFormat: GeneratedContentFormat | undefined;
+  answer: string | undefined;
+  citations: GeneratedAnswerCitation[] | undefined;
+  generated: boolean;
+  isStreaming: boolean;
+  isLoading: boolean;
+}
+
+interface Data {
+  answerStyle: string | undefined;
+  contentFormat: GeneratedContentFormat | undefined;
+  answer: string | undefined;
+  citations: GeneratedAnswerCitation[] | undefined;
+  generated: boolean;
+  isStreaming: boolean;
+  isLoading: boolean;
+}
+
+const updateCacheWithEvent = (event: EventSourceMessage, draft: Data) => {
+  const message = JSON.parse(event.data);
+  const payload = JSON.parse(message.payload);
+  if (message.payloadType === 'genqa.headerMessageType') {
+    draft.answerStyle = payload.answerStyle;
+    draft.contentFormat = payload.contentFormat;
+    draft.isStreaming = true;
+    draft.isLoading = false;
+  }
+
+  if (message.payloadType === 'genqa.messageType') {
+    if (draft.answer === undefined) {
+      draft.answer = payload.textDelta;
+    } else {
+      draft.answer = draft.answer?.concat(payload.textDelta);
+    }
+  }
+
+  if (message.payloadType === 'genqa.citationsType') {
+    draft.citations = payload.citations;
+  }
+
+  if (message.payloadType === 'genqa.endOfStreamType') {
+    draft.generated = Boolean(payload.answerGenerated);
+    draft.isStreaming = false;
+  }
+};
+
+const onOpenStream = async (response: Response) => {
+  if (
+    response.ok &&
+    response.headers.get('content-type')?.includes('text/event-stream')
+  ) {
+    return;
+  }
+
+  const isClientSideError =
+    response.status >= 400 && response.status < 500 && response.status !== 429;
+
+  if (isClientSideError) {
+    throw new FatalError({
+      message: 'Error opening stream',
+      code: response.status,
+    });
+  } else {
+    throw new Error();
+  }
+};
+
+const onError = (err: Error) => {
+  let retryCount = 0;
+
+  if (err instanceof FatalError) {
+    throw err;
+  }
+
+  if (++retryCount > 3) {
```

Comment:
if we set the retryCount on line 97, how can this ever be true ?

Reply:
Good question. It will never be true even with just 3.  Since the number of retry is actually limited to 3 directly at the root of the slice.
https://github.com/coveo/ui-kit/pull/4014/files#diff-2a2048d0918dc62fa9b3767828d61914422ff71fc7d8b395cf163c2b0afc1e19R42

I will remove the check here.